### ISSUE TITLE
Implementing an object manager

### ISSUE THREAD

**Issue by RomainLaniece** [OP-TEE/optee_os#3084]
Hi,

I'm currently trying to identify how I could implement a kind of object manager TA, which would be in charge of storing/opening my persistent objects. I saw that the idea had already been given here [issue_1068](https://github.com/OP-TEE/optee_os/issues/1068#issuecomment-249131741), and I was confident in being successful.

But I faced the transient object : my idea was to create transient object into a TA, give it to my object manager (while invoking a command, with at least one parameter containing params[x].memref.buffer = &object), so that this one would be able to store the persistent object based on the transient one from the original TA.

The issue I faced is the using of this pointer on the other side (into the object manager).
Actually, I can e.g. call TEE_GetObjectInfo1() with success by using the pointer of the transient object into the original TA. But once into the object manager, calling TEE_GetObjectInfo1() returns me an error (0xFFFF0006) and the info structure is empty.

I'm quite sure that the issue is because the transient object can only be used in one TA environment exclusively, but I didn't find this information in any documentation, can someone confirm it ?
If yes, then how can I implement such a manager ? Is creating both the transient and the persistent object into the same TA the solution ?

Best Regards.

PS : I currently don't have the hand on the op-tee version, and it seems I'm using the version 1.1

**Comment by jbech-linaro** [comment]
I'm don't think I follow along with what you are trying to achieve (and I have re-read #1068). Is it #1068 that is you use case? If not, would you mind elaborate a bit more?

> PS : I currently don't have the hand on the op-tee version, and it seems I'm using the version 1.1

You must be a lonely guy in that case. That release is ~5 years old. That is probably a greater problem than implementing an object manager ;-)

**Comment by RomainLaniece** [comment]
Hi,

Thanks for the answer ^^

According to the version I use, this is the one I have in the SDK I received from a chip founder, and I have no way to update it right now, but I'm currently not sure that the version is the reason of the problem I'm facing up to (but I may be wrong).

I already faced the fact that it was not possible to share objects between TA. By looking for an idea, I've read the [#1068](https://github.com/OP-TEE/optee_os/issues/1068) and there was the advice to implement one TA to manage the object for others TA. But my first issue is the way I create the object.

Actually, if one TA need to create an object, then it has to share a kind of template to the TA that will create the persistent object. I thought I would be able to share the transient as such a template. It appears that I couldn't, which is the reason I created this issue. And I was wondering if in the end it was possible. Here is a kind of summary of the use case:

1. regular TA wants to create an object e.g. RSA keypair
2. regular TA prepares a template (transient object ?)
3. regular TA shares the template with the object manager TA
4. object manager TA creates a persistent object from this template -> issue as the template is empty/not shareable

The reason of such an object manager would be to avoid the copy/paste of the storage functions I would implement into all of my TA, so that I could share the source code between those TA.

Now that I've faced this issue, I'm also wondering how I could restore/share the content of an object to a TA from that TA/ObjectManager, because I don't know how to dump an object in order to share the data, but I guess this is the purpose of another issue.

Best Regards
